File: api/changelog.py
import bs4, requests
from http.server import BaseHTTPRequestHandler
from urllib.parse import urlparse, parse_qs


def scrape(id):
  url = f'https://routinehub.co/shortcut/{id}/changelog'
  req = requests.get(url)
  req.raise_for_status()
  soup = bs4.BeautifulSoup(req.text, 'html.parser')
  return soup

def extract(soup, selector):
  elems = soup.select(selector)
  return elems


class handler(BaseHTTPRequestHandler):
  def do_GET(self):
    parsed_path = urlparse(self.path)
    path = '?' + parsed_path.query
    try:
      RoutineHubID =  parse_qs(path[1:])["id"][0]
      isValid = True
      soup = scrape(RoutineHubID)
      if 'Error: Shortcut not found' in str(soup):
        isValid = False
        data = 'The provided id does not exist or it\'s invalid.'
    except:
      isValid = False
      data = 'Required parameter was not given or was incorrect. Check docs at https://rh-api.alombi.xyz'
    
    if isValid:
      text = extract(soup, '#content > div > div.versions')

      data = {
        "id":RoutineHubID
      }
      elems = extract(soup, '#content > div > div.heading > h4')
      res = elems[0].text.strip()
      data["name"] = res
      # Get number of versions
      number_of_versions = 0
      for word in str(text).split():
        if word == '<article':
          number_of_versions = number_of_versions + 1
      data["updates"] = number_of_versions
      child = 1
      i = number_of_versions
      versions = {
        "versions":[]
      }
      while i != 0:
        text = extract(soup, f'#content > div > div.versions > article:nth-child({str(child)})')
        version = str(text).split('\n')[4].split('</strong>')[0].replace('<strong>', '')
        versionData = {
          "version":version
        }
        release_date = str(text).split('\n')[4].split('<small>')[1].replace('</small>', '')
        versionData["release_date"] = release_date
        iOS = str(text).split('\n')[6].replace('</small>', '').replace('<small>', '')
        versionData["iOS"] = iOS
        release_notes = str(text).split('<br/>')[2].split('<br/>')[0]
        versionData["release_notes"] = release_notes
        downloads = str(text).split('\n')[10].replace('</em>', '').replace('<em>', '')
        versionData["downloads"] = downloads

        versions["versions"].append(versionData)

        child = child + 1
        i = i -1 
      data["versions"] = versions
      data = str(data).replace('\'', '\"')

    self.send_response(200)
    self.send_header('Content-type', 'text/plain')
    self.end_headers()
    self.wfile.write(data.encode())
    return

File: api/changelog.ts
import puppeteer from 'puppeteer';

async function checkForResults(page){
    let notFoundString = await page.waitForSelector('#content > div > div > div');
    notFoundString = await notFoundString?.evaluate((e) => e.innerText);
    if(notFoundString == 'No results found.'){
        return true;
    }else{
        return false;
    }
}

module.exports = async (req, res) => {
    const query: string = req.query.q;
    const baseURL: string = `https://routinehub.co/search/?q=${query}`;

    const browser = await puppeteer.launch();
    const page = await browser.newPage();
    page.setUserAgent('Mozilla/5.0 (iPhone; U; CPU iPhone OS 3_0 like Mac OS X; en-us) AppleWebKit/528.18 (KHTML, like Gecko) Version/4.0 Mobile/7A341 Safari/528.16')
    await page.goto(baseURL);

    const hasNoResults = await checkForResults(page);
    if(hasNoResults){
        res.json({
            results: [],
            error: 'No results found.'
        })
    }
    // const results = await page.evaluate(() =>
    //     Array.from(document.querySelectorAll('a'),(e)=>({
    //         id: Number(e.querySelector('a')?.href)
    //     }))
    // )
    let container = await page.waitForSelector('#content > div > div');
    const results = await container?.evaluate((e) => Array.from(e.querySelectorAll('.column'), (el) => ({
        name:el.querySelector('strong')?.innerText,
        id: Number(el.querySelector('a')?.href.replace('https://routinehub.co/shortcut/', '').replace('/', '')),
        description: el.querySelector('small')?.innerText,
        downloads: Number(el.querySelectorAll('small')[1]?.innerText),
        hearts: Number(el.querySelectorAll('small')[2]?.innerText),
        link: el.querySelector('a')?.href,
        api_link: 'https://rh-api.alombi.xyz/shortcut?id=' + Number(el.querySelector('a')?.href.replace('https://routinehub.co/shortcut/', '').replace('/', '')),
        routinehub_api_link: `https://routinehub.co/api/v1/shortcuts/${Number(el.querySelector('a')?.href.replace('https://routinehub.co/shortcut/', '').replace('/', ''))}/versions/latest`
    })))
    const totalResults = results?.length;
    

    res.json({
        totalResults: totalResults,
        results:results
    })
}